### ISSUE TITLE
Add a version number to the library + git tag

### ISSUE THREAD

**Issue by aurelienrb** [abseil/abseil-cpp#87]
Hello,

I can't see any version number / scheme in the library. Does it exist?
It would be nice to have such version numbers reflected via git tags (and probably the cmake file) in order to:
- clone the repository on a specific tag in order to ensure repeatable builds & environment setup
- know if a new version has been released and compare what's new regarding the version we are currently using

That would probably imply to maintain a changelog file as well. It's a bit of more work but it's very useful for consumers like me.

Aurelien

**Comment by aurelienrb** [comment]
Note that there's a similar issue with the abseil dependency `cctz` : the last tag is `v2.1` but it does not contain the CMakeLists.txt file that exists in master.

**Comment by ahedberg** [comment]
Hi Aurelien,

Abseil is not using semantic versioning. We're recommending that users choose to [live at head](https://abseil.io/about/philosophy#we-recommend-that-you-choose-to-live-at-head). If specifying a specific commit is important for repeatable builds, we encourage folks to update their Abseil dependency regularly.

Titus Winters' [CppCon talk](https://www.youtube.com/watch?v=tISy7EJQPzI) describes the pitfalls with semantic versioning and goes into more detail about C++ as a live-at-head language.

**Comment by aurelienrb** [comment]
Hello,

I think 2 things should be distinguished:
- dependency management through semantic versioning
- traceability management through any kind of versioning (release date...) + changelog

I understand the issues with the first point, but I am dealing with the second one. For example, I can be forced for legal reasons to be able to list the complete and exact code that has been compiled and shipped into my medical device 5 years ago. I also want to know that the version I shipped 2 years ago actually contains a bug / security issue that has been fixed recently.

So my question is: how to (easily) deal with such aspects without a changelog + versioning scheme ? Note that it does not have to be SemVer, it can simply be a git tag with the release date (like Ubuntu) or a more abstract number (like Chrome).

>  we encourage folks to update their Abseil dependency regularly

Yep, that's actually why I need a version number: to detect a new release is available and push into to a package management system. I will probably use the release (commit) date as a version number. But there's still the issue of not being able to tell what's new in this release compared to the previous ones.

Note that being able to do that goes in the direction you want to promote: update regularly by being notified a new version is out. Otherwise, most people just copy&paste the code in their repo and never update it.

Do you have an opinion on how to ship Abseil as a package?

Aurelien

**Comment by ahedberg** [comment]
Our commit messages contain a log of everything that changed in that commit, e.g. [this one](https://github.com/abseil/abseil-cpp/commit/0ec11bad6fc18822e851b25344f027491cc85746) from 2 days ago. We push a new version roughly every US business day. We're concerned that having a changelog beyond what we provide in commit messages will affect whether users update their versions or not--and we really want users to live at head. Until we have an automated solution for living at head (which we're actively working on), we're advising folks to update the commit tag they're using on a regular basis. Can you elaborate on why using the head commit from the master branch won't work for you? I'm not understanding why a number beyond the commit hash is important.

Abseil's philosophy is that users will build from source, so we're not investigating shipping Abseil as a package.

**Comment by aurelienrb** [comment]
> Our commit messages contain a log of everything that changed in that commit

Indeed, but that's considered bad practice:
http://keepachangelog.com/en/1.0.0/#bad-practices

I started to use abseil a few months ago, and I have no idea of what has been added since then. Glancing the code base or a noisy git log is painful. As a result, I'm less likely to use more features and benefit from frequent updates.
By *frequently*, what do you mean exactly? Is it once a day/week/month/quarter?

> We're concerned that having a changelog beyond what we provide in commit messages will affect whether users update their versions or not

It's a trade-off between how many people will not update vs how many will do it more frequently if you provide a changelog. IMO, providing your users a convenient way to learn about the new features will encourage them to update. OTOH, they have the right to consider that the changes are not interesting for them and deny the update. That's their responsibility... and freedom.

Many companies develop a single monolithic app. They never suffer from the diamond issue and are quite happy to update once or twice a year their libs - if they do. When they do, they want to know what to expect from the update. Having a changelog helps a lot to plan and communicate about that operation. Unpredictable changes tend to be delayed and avoided.

> Can you elaborate on why using the head commit from the master branch won't work for you?

There are various reasons. The first one is not working with git. Many projects are using svn or another VCS. In the industry, that's very common. So there's no master head available.

Even if you work with git, you might still prefer to copy-paste (vendor) the dependency into your source tree:
https://github.com/mongodb/mongo/tree/master/src/third_party

Then updating by hand becomes a pain. And a package manager becomes a good option. Note that a library can be packaged as plain source files.

> I'm not understanding why a number beyond the commit hash is important.

It's a matter of being human friendly. If you don't maintain a version number, then the commit hash becomes the version number. Since it's quite random, it's much more difficult to reason with than an incremental version scheme that gives an idea of the distance between 2 versions.

---

Actually I am working on a toolbox of good tools and libraries to help people discover and adopt better C++ practices. Basically they will download or receive (during a training) a ready to use project template and built on top of it. I'm trying to add abseil to that toolbox. But I face the issue of versioning and documenting my toolbox changes with a library that has no version / changelog.

**Comment by ahedberg** [comment]
It sounds like we're coming at this from different premises. We know live-at-head is a big jump, and we'd like to hear more about where you're coming from. If you're interested in a more real-time method of communication, feel free to send me an email (append "@google.com" to my username), and we'll coordinate from there.

**Comment by LebedevRI** [comment]
I'd like to re-surface this issue. I'm analyzing the possibility of using abseil in https://github.com/darktable-org/rawspeed/issues/122

https://abseil.io/about/philosophy#if-you-choose-to-not-live-at-head states:
```
If You Choose to Not Live at Head:
* You are of course free to fork Abseil, include it in your project and manage dependencies any way that is convenient for your development practices.
* We will periodically mark a tag as “supported” and branch it — if we discover security issues or major performance problems, we’ll update those branches. Our expectation is to do this every 6 months and support those branches for 2 years.
```

(As of right now, there are no tags, and no versioned branches in the repo.)
Two questions:
* Is that still true?
* If yes, when is the first tag planned?

**Comment by ahedberg** [comment]
I expect that the Abseil team will discuss versioned branches and tags this week. We don't have an answer for you yet, but should in a couple of days.

**Comment by isaachier** [comment]
@ahedberg any news yet?

**Comment by ahedberg** [comment]
Some more internal discussions need to happen first, but:

- Yes, we still plan to periodically mark a tag as “supported” and branch it.
- We plan to make the first one soon, hopefully within the next few weeks.

**Comment by isaachier** [comment]
You could do what [folly](github.com/facebook/folly/releases) does and use the current date. Just need a marker to download the same copy from github.

**Comment by LebedevRI** [comment]
@ahedberg has there been talks with the actual distro people (packagers), who know how linux packaging actually works?
As highlighted in https://github.com/darktable-org/rawspeed/issues/122#issuecomment-370947396, there **may** be some rather noticeable flaws with the current design. It might be best to at least talk about it.

**Comment by aurelienrb** [comment]
May be you should also talk with the folks developing Go. Basically they followed a "live at head" strategy for many years with the `go get` command that can only fetch the latest version available on the master branch of a given dependency. And now we know this is problematic.

That's why they introduced support for source code vendoring and are working hard today on fixing this "live at head" issue:
https://golang.github.io/dep/docs/FAQ.html#does-dep-replace-go-get

After all, git is a *version* control system, it is meant to control versions :)

**Comment by isaachier** [comment]
Too bad C++ doesn't allow for multiple definitions of the same symbol (one definition rule). Go can take advantage of the fact that packages are more or less equivalent to the basename of the directory containing the source. Namespaces in C++ have no relationship to the directory so symbols cannot be adjusted accordingly.

**Comment by aurelienrb** [comment]
Actually C++ allows for multiple definitions across different translation units for inline / template code. The constraint is they must be consistent. I wonder how a language could not follow that principle, but in C/C++ it's easy to link modules built with different compiler / preprocessor flags, thus the deception.

Go has its own package versioning strategy by playing on the folder names : mypackage-v1, mypackage-v2, etc... Their initial "live at head" vision was that a git repository should be used to maintain a stable API. And if you need to introduce breaking changes, just create a new repository for that new API...
http://zduck.com/2014/go-and-package-versioning/
IMO we loose all the benefits of using a tool such as git. And it seems the Go team has realized it too.

There are many real life situations where multiple versions of the same software *have to* be maintained in parallel for years. And where semantic versioning with the concept of minor / major / breaking change works well (Qt for example). Proposing an alternative is great. Forcing to use it creates resistance. Resistance to adopt abseil.

**Comment by LebedevRI** [comment]
> There are many real life situations where multiple versions of the same software have to be maintained in parallel for years.

Actually, it's even worse than that.
And from what is known so far, i'm not too sure the actual problem has been considered.

**Comment by isaachier** [comment]
@aurelienrb sorry for the late reply. I think C++ gets away with it with templates because of name mangling. And by inline, at least in theory, there shouldn't be a symbol emitted anyway. But consistently defining symbols that are redefined based on the location of the object is not part of the C/C++ repertoire. Thus, we are forced to use only one version of any given library.

**Comment by aurelienrb** [comment]
@isaachier *Thus, we are forced to use only one version of any given library.* Ah ok, I get your point now, thanks for the clarification! Regarding the possibility to have multiple definitions of a symbol, there's a good clarification [here](https://stackoverflow.com/a/4193698).

**Comment by LebedevRI** [comment]
> Thus, we are forced to use only one version of any given library.

@JonathanDCohen, @ahedberg
The question that needs to be answered along with deciding on tagging is:
* If only the program uses abseil, then it can use one version of it.
* If the program also uses a library that uses abseil, then it needs to use the same version of abseil as that library. Cool.
* But if the program uses two or more libraries that use abseil, then the clusterfuck begins.

There are solutions:
* The ignorant one: "you can't use abseil in proper libraries that aren't copylibs/header-only / do proper releases".
   But then
   1. you won't see wide adoption of abseil
   2. someone somewhere will use it in such a library anyway.
* The invalid one: "just make sure that everything is always built with abseil trunk."
   Hint: unless every one of those libraries is supposed to be built in-tree (which the packagers will *hate* with flaming passion, rightfully), it won't happen/help.
   Because even if the abseil trunk was used when each of those libraries was packaged, nothing prevents them from being packaged at different times, thus using different abseil HEAD.
* The sane-ish solution:
   1. Do create a tag/release every `n` months.
   2. Tell distros to package one and only one abseil version - that last tag.
   3. Tell distros/packagers that one and only way the packaged programs can use abseil is by using that distro-provided abseil.
       * The abseil shipped with the software **MUST** be replaced with the abseil version from the package at distro packaging time.
       * *Every* new version of that [source] package **MUST** be an equivalent of ABI version bump.
       * No two versions must be co-installable.
       * Every new version of that package **MUST** cause rebuilds of all the software that uses it.
   
   This will, of course, cause troubles once an API-incompatible change happens in abseil, that requires changing code in order to upgrade.

**This** is the problem. This is the problem that *needs* to be addressed.
I'm not seeing any other solution. Please tell me i'm wrong :)

**Comment by isaachier** [comment]
@LebedevRI very good summary of the problem! I don't think Google cares to package any library that depends on Abseil. The reasoning: why else use Bazel to clone everything from the web if you already have packages installed? Implication is that Google internally has no use for the distros so they aren't relying on them at all.

**Comment by aurelienrb** [comment]
> This will, of course, cause troubles once an API-incompatible change happens in abseil, that requires changing code in order to upgrade.

AFAIK, the position of Google / Titus is that every change is potentially a breaking change. Therefore semver is a fallacy (looking at how many npm packages fail to respect the contract with a language much simpler than C++ seems to prove him right).

My position is that source code versioning is not just about managing breaking changes, it's also about managing features.

**Comment by JonathanDCohen** [comment]
> AFAIK, the position of Google / Titus is that every change is potentially a breaking change

Yes, this is the position of Abseil.  We will never use semantic versioning.

As for releases, we discussed this a good amount at the summit last week.  Here is a dump of thoughts --

We will begin our Long Term Support tags once two things happen -- source installation with CMake, which I'm currently writing, and a story for an internal issue which should resolve in the next couple of weeks.  A Long Term Support release will simply be a tagged commit which we will backport critical / security fixes for two years.  We plan to cut these releases around every 6 months.

What we are still holding on is not providing any kind of binary version of Abseil.

I also was wondering about tagging our daily commit as a new release as well, so that package manager maintainers can also provide Abseil HEAD.  How does everyone feel about this idea?

**Comment by isaachier** [comment]
No offense, but cmake installation is something almost every project does wrong. I blame the CMake docs for hiding this essential step deep in their site. Here is an excellent sample I base all of my CMake installations on: https://github.com/forexample/package-example/blob/master/Foo/CMakeLists.txt.

**Comment by JonathanDCohen** [comment]
> No offense, but cmake installation is something almost every project does wrong.

Can you be more specific?

**Comment by isaachier** [comment]
A few things. First and foremost they don't install a package config. Not to be confused with pkg-config, CMake package configs export the targets for other CMake projects. Then, if they do somehow install a config, they will hand write ugly variables such as ABSEIL_LIBRARIES and ABSEIL_INCLUDE_DIRECTORIES. this pattern is no longer helpful to users. If you use the install(export) mechanism and you install a package config, I can do `target_link_libraries(myproject abseil::abseil)` and I automatically pull in all the include flags and compile flags. This is why I highly recommend you look at that example please so I don't end up rewriting it like I did for cctz and a dozen others... See https://github.com/hunter-packages/cctz for example.

**Comment by JonathanDCohen** [comment]
Yes, the plan is to do an install(export) to install abseil CMake targets in the absl:: CMake namespace

**Comment by LebedevRI** [comment]
> source installation with CMake, which I'm currently writing

Ok, now i'm fully lost the though. Why install sources?
I guess there will be **zero** reaction to the points in https://github.com/abseil/abseil-cpp/issues/87#issuecomment-372945260. Sad.

**Comment by JonathanDCohen** [comment]
I hear your frustration, and we're trying to come up with solutions which work for everyone.  Our plan is to do tagged releases every 6 months, and possibly also tag our daily releases.  This sounds more or less like your "sane-ish" plan.  Is there something else you're looking for?

**Comment by LebedevRI** [comment]
Yes, adding a warning page to the website that
1. Will warn distros about The Issue (`But if the program uses two or more libraries that use abseil, then the clusterfuck begins.`),
2. And _suggest_ the way forward (see `.. to package one and only one abseil version - that last tag.` + `.. that one and only way the packaged programs can use abseil is by using that distro-provided abseil.`)

I have no idea if this will work out in practice, but i can **guarantee** you from practice that if it is *NOT* done, there **will** be issues, as already previously seen in the wild with LLVM:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=846410
https://bugs.kde.org/show_bug.cgi?id=373614

**Comment by perfinion** [comment]
I just want to chime in too and ask for version numbers. I have been trying to package all the dependencies of TensorFlow on Gentoo Linux and this is one of the few that I cant do because there is no release.

I understand google internally just uses HEAD but distros don't work like that. the package manager needs to compare numbers, v1.2 is older than v2.5 is easy to figure out. Is 9acad869d2 older or newer than 3c491f6fd? I have no idea.

It doesn't need to be full semver, just any monotonically increasing number would help immensely. The date is a pretty good format for projects that don't want to deal with semver. 20180801 could be a release now, then next month's release is 20180901. or once a quarter or once a year or however often you'd like to support new versions.

> I also was wondering about tagging our daily commit as a new release as well, so that package manager maintainers can also provide Abseil HEAD. How does everyone feel about this idea?

That is a bit much. I won't be able to bump abseil every single day, and I definitely wouldn't want to make users build a package every single day either. Also the version numbers would be huge really quick. Once a month is probably plenty often enough. If users ask for it, we sometimes package a "live" version which will install the latest from HEAD but that is a special case and for users that actively develop with the project.

> Yes, adding a warning page to the website that
> Will warn distros about The Issue (But if the program uses two or more libraries that use abseil, then the clusterfuck begins.),
> And suggest the way forward (see .. to package one and only one abseil version - that last tag. + .. that one and only way the packaged programs can use abseil is by using that distro-provided abseil.)
> I have no idea if this will work out in practice, but i can guarantee you from practice that if it is NOT done, there will be issues, as already previously seen in the wild with LLVM:

That is standard practice for distros. Packages depend on the versions of libraries they require and the package manager figures it out. There is nothing unique about abseil in this regard, if two conflicting versions of libpng were used the same would happen. As the upstream project this is not something you'd have to worry about, distros already have to handle this. Generally the problem is worse if packages bundle their deps (which is why I've been working to unbundle the deps in TensorFlow)

**Comment by tituswinters** [comment]
On Wed, Aug 1, 2018 at 9:29 AM Jason Zaman <notifications@github.com> wrote:

> I just want to chime in too and ask for version numbers. I have been
> trying to package all the dependencies of TensorFlow on Gentoo Linux and
> this is one of the few that I cant do because there is no release.
>
> I understand google internally just uses HEAD but distros don't work like
> that. the package manager needs to compare numbers, v1.2 is older than v2.5
> is easy to figure out. Is 9acad86
> <https://github.com/abseil/abseil-cpp/commit/9acad869d21731f5bc50430a33fe61cc0ffcbb0b>
> older or newer than 3c491f6
> <https://github.com/abseil/abseil-cpp/commit/3c491f6fd0f4cdb0f5772ad8ec63b36dea7225a7>?
> I have no idea.
>
> It doesn't need to be full semver, just any monotonically increasing
> number would help immensely. The date is a pretty good format for projects
> that don't want to deal with semver. 20180801 could be a release now, then
> next month's release is 20180901. or once a quarter or once a year or
> however often you'd like to support new versions.
>
We're planning / supporting LTS branches for this purpose:
https://abseil.io/blog/20180618-lts-branches

Does that suffice?

**Comment by perfinion** [comment]
Cool, thats helpful.
How do I number the packages then?
For example: https://packages.gentoo.org/packages/dev-python/absl-py
I have packaged absl-py-0.2.2 (i just saw 0.3.0 is out, i'll bump that soon)
I can package abseil-cpp-20180620, are there going to be tags too?
What about when new commits are pushed to the branch? what should the version be then?

The package called "absil-cpp-20180620" must be completely reproducible, package managers check the hashes of all the tarballs used and then builds in a sandbox with no internet access. I can't change what a package points to once its out. Are there going to be eg 20180620.1 releases too?

**Comment by mbxx** [comment]
I'd recommend numbering the LTS branches YYYYMMPP (Year,Month,Patch), which in this case would be 20180600.  That's the format we've talked about using internally.  This seems reasonable to us because the LTS branches will be cut somewhat infrequently (every 6 months or so) and while branches may receive some commits, we expect those to be very rare.  The only commits will be critical bug fixes (think fixing a security vulnerability), and I haven't seen anything that would qualify for one in the two years I've been working on this project.

**Comment by perfinion** [comment]
> I'd recommend numbering the LTS branches YYYYMMPP (Year,Month,Patch), which in this case would be 20180600.

That sounds reasonable. Maybe with dots between the parts? Either way can I ask you guys to push some git tags explicitly on those commits? Having the tags makes github generate tarballs with names that are easier to fetch and would take the guess work out of coordinating between distros and people consuming it. I don't want one distro's "201806.01" to be the same as another distros "201806.02" just because they counted differently. It would make bug reports much more standardized.

**Comment by LebedevRI** [comment]
Does "LTS branch" here mean API/ABI-stable?

**Comment by tituswinters** [comment]
@LebedevRI - As stable as we can make it. Practically speaking, the LTS branches are going to be tags in every case we've really imagined.  However, there is always the possibility that there are security issues that arise (I'm not entirely sure what those would look like at this low a level in the stack). If something interesting / important comes up, we'll backport fixes to extant LTS branches going back a couple years - and in those cases we'll try to keep it as API and even ABI stable as possible.  The abstract nature of "security fix" doesn't always allow for such.

**Comment by mbxx** [comment]
@perfinion: Adding git tags to the LTS branches sounds reasonable.  We've gone ahead and tagged the first LTS.

I think that covers what's been requested, so I'm going to close this issue out.  Please let us know if this doesn't meet your needs for any reason.
